src/upsert.rs

```diff
@@ -68,60 +72,79 @@ pub trait FindexUpsert<
     /// - `additions`   : values to indexed for a set of keywords
     /// - `deletions`   : values to remove from the indexes for a set of
     ///   keywords
+    ///
+    /// Returns a map of keywords to booleans indicating whether the keyword
+    /// was already present in the database.
     async fn upsert(
         &mut self,
         master_key: &KeyingMaterial<MASTER_KEY_LENGTH>,
         label: &Label,
         additions: HashMap<IndexedValue, HashSet<Keyword>>,
         deletions: HashMap<IndexedValue, HashSet<Keyword>>,
-    ) -> Result<(), Error<CustomError>> {
+    ) -> Result<HashMap<Keyword, bool>, Error<CustomError>> {
         check_parameter_constraints::<CHAIN_TABLE_WIDTH, BLOCK_LENGTH>();
 
         let mut rng = CsRng::from_entropy();
         let k_uid = master_key.derive_kmac_key::<KMAC_KEY_LENGTH>(ENTRY_TABLE_KEY_DERIVATION_INFO);
         let k_value = master_key.derive_dem_key(ENTRY_TABLE_KEY_DERIVATION_INFO);
 
-        let mut new_chains =
-            HashMap::<KeywordHash, HashMap<IndexedValue, BlockType>>::with_capacity(
-                additions.len() + deletions.len(),
-            );
+        // a reverse map of Keyword UIDs (e.g. hashes) to Keywords
+        let mut uid_to_keyword = HashMap::<Uid<UID_LENGTH>, Keyword>::new();
+
+        let mut new_chains = HashMap::<Keyword, HashMap<IndexedValue, BlockType>>::with_capacity(
+            additions.len() + deletions.len(),
+        );
         for (indexed_value, keywords) in additions {
             for keyword in keywords {
                 new_chains
-                    .entry(keyword.hash())
+                    .entry(keyword)
                     .or_default()
                     .insert(indexed_value.clone(), BlockType::Addition);
             }
         }
         for (indexed_value, keywords) in deletions {
             for keyword in keywords {
                 new_chains
-                    .entry(keyword.hash())
+                    .entry(keyword)
                     .or_default()
                     .insert(indexed_value.clone(), BlockType::Deletion);
             }
         }
         // Compute the Entry Table UIDs.
         let mut new_chains = new_chains
             .into_iter()
-            .map(|(keyword_hash, indexed_values)| {
-                (
-                    EntryTable::<UID_LENGTH, KWI_LENGTH>::generate_uid::<KMAC_KEY_LENGTH>(
-                        &k_uid,
-                        &keyword_hash,
-                        label,
-                    ),
-                    (keyword_hash, indexed_values),
-                )
+            .map(|(keyword, indexed_values)| {
+                let keyword_hash = keyword.hash();
+                let uid = EntryTable::<UID_LENGTH, KWI_LENGTH>::generate_uid::<KMAC_KEY_LENGTH>(
+                    &k_uid,
+                    &keyword_hash,
+                    label,
+                );
+                uid_to_keyword.insert(uid, keyword);
```

Comment:
Do not add side effects to the maps. You could implement this using `unzip()` if you want to keep the functional style. Otherwise just use a `for` loop.

Reply:
let us go for a for loop; it is going to be hard to read with an unzip